### PR TITLE
[11.0][MIG] account_reconcile_rule (account_operation_rule v9)

### PR DESCRIPTION
**Pull request by grindtildeath**
Todo :
 - [x] Fix Javascript
 - [x] Check readme

### CODE REVIEW

File: account_reconcile_rule/static/src/js/account_widgets.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+//                if (self.preset_auto_clicked) {
+//                    self.addLineBeingEdited();
+//                }
+//            });
+//        },
+//        restart: function () {
+//            var deferred = this._super();
+//            deferred.done(this.operation_rules());
+//            return deferred;
+//        }
```
Comment (guewen) [inline account_reconcile_rule/static/src/js/account_widgets.js]: I guess the commented lines can be removed

File: account_reconcile_rule/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Author: Guewen Baconnier
+# © 2014-2016 Camptocamp SA
```
Comment (simahawk) [inline account_reconcile_rule/__manifest__.py]: `Copyright`

File: account_reconcile_rule/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+{'name': 'Account Reconcile Rules',
+ 'version': '11.0.1.0.0',
+ 'author': 'Camptocamp, Odoo Community Association (OCA)',
+ 'maintainer': 'Odoo Community Association (OCA)',
+ 'license': 'AGPL-3',
+ 'category': 'Accounting & Finance',
+ 'depends': [
+     'account',
+ ],
+ 'website': 'http://www.camptocamp.com',
```
Comment (simahawk) [inline account_reconcile_rule/__manifest__.py]: https://github.com/OCA/account-reconcile

File: account_reconcile_rule/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ 'data': [
+     'view/account_reconcile_rule.xml',
+     'view/account_reconcile_rule_view.xml',
+     'security/ir.model.access.csv',
+ ],
+ 'installable': True,
```
Comment (simahawk) [inline account_reconcile_rule/__manifest__.py]: you can drop these flags

File: account_reconcile_rule/i18n/account_statement_operation_rule.pot [hunk 5]
```diff
@@ -0,0 +1,152 @@
+# Translation of Odoo Server.
```
Comment (simahawk) [inline account_reconcile_rule/i18n/account_statement_operation_rule.pot]: remove POT

File: account_reconcile_rule/model/__init__.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (simahawk) [inline account_reconcile_rule/model/__init__.py]: empty line not needed

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    rule_type = fields.Selection(
+        selection=[('rounding', 'Roundings'),
+                   ('currency', 'Currencies')],
+        string='Type',
+        default='rounding',
+        required=True,
+    )
+    reconcile_model_ids = fields.Many2many(
+        comodel_name='account.reconcile.model',
+        string='Operations'
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: trailing comma

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        if currency.compare_amounts(value, low) == -1:
+            return False
+        if currency.compare_amounts(value, high) == 1:
+            return False
+        return True
+
+    @api.multi
+    def _balance_in_range(self, balance, currency):
+        amount_min = self.amount_min
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: you don't need these vars

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    @api.multi
+    def _is_valid_balance(self, statement_line, balance):
+        if self._is_multicurrency(statement_line):
+            return False
+        currency = statement_line.currency_for_rules()
+        return self._balance_in_range(balance, currency)
+
+    @api.multi
+    def _is_valid_multicurrency(self, statement_line, move_lines, balance):
+        """ Check if the multi-currency rule can be applied
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: no wrapping space

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+        # amount in currency is the same, so the balance is
+        # a difference due to currency rates
+        if statement_line.currency_id.is_zero(amount_currency):
+            return self._balance_in_range(balance, currency)
+        return False
+
+    @api.multi
+    def is_valid(self, statement_line, move_lines, balance):
+        """ Returns True if a rule applies to a group of statement_line +
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: `"""Check if a rule applies to a group of statement_line + move lines.`

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        if self.rule_type == 'rounding':
+            return self._is_valid_balance(statement_line, balance)
+        elif self.rule_type == 'currency':
+            return self._is_valid_multicurrency(statement_line,
+                                                move_lines,
+                                                balance)
+
+    @api.model
+    def find_first_rule(self, statement_line, move_lines):
+        """ Find the rules that apply to a statement line and
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: `"""Find rules to apply to given statement line and move lines.`

File: account_reconcile_rule/model/account_reconcile_rule.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        # return the first applicable rule
+        for rule in rules:
+            if rule.is_valid(statement_line, move_lines, balance):
+                return rule
+        return self.browse()
+
+    @api.model
+    @api.returns('account.reconcile.model')
+    def models_for_reconciliation(self, statement_line_id, move_line_ids):
+        """ Find the rule for the current reconciliation and returns the
```
Comment (simahawk) [inline account_reconcile_rule/model/account_reconcile_rule.py]: `"""Find reconcile models for given statement and move lines.`
explain "how" in the extra lines.

File: account_reconcile_rule/static/src/js/account_widgets.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
+odoo.define('account_reconcile_rule', function (require) {
+    "use strict";
+
+    var core = require('web.core');
+    var reconciliation_renderer = require('account.ReconciliationRenderer');
+
```
Comment (simahawk) [inline account_reconcile_rule/static/src/js/account_widgets.js:6]:
```javascript\
/*
Explain what you do here 
*/`
```

File: account_reconcile_rule/static/src/js/account_widgets.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        start: function () {
+            var self = this;
+            var deferred = this._super();
+            if (deferred) {
+                deferred.done(this.reconciliation_rule_models());
+            }
+            return deferred;
+        },
+        /**
+         * Click on add_line link if preset button have been clicked
```
Comment (simahawk) [inline account_reconcile_rule/static/src/js/account_widgets.js]: ???

File: account_reconcile_rule/tests/common.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    def setUpClass(cls):
+        super(AccountReconciliationModelTestCase, cls).setUpClass()
+        cls.reconcile_model_obj = cls.env['account.reconcile.model']
+        cls.rule_obj = cls.env['account.reconcile.rule']
+        cls.cash_journal = cls.env['account.journal'].create({
+            'name': 'Unittest Cash journal',
+            'code': 'CASH',
+            'type': 'cash',
+        })
+
```
Comment (simahawk) [inline account_reconcile_rule/tests/common.py]: remove some useless empty line

File: account_reconcile_rule/tests/common.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests.common import SavepointCase
+
+
+class AccountReconciliationModelTestCase(SavepointCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super(AccountReconciliationModelTestCase, cls).setUpClass()
```
Comment (simahawk) [inline account_reconcile_rule/tests/common.py]: you can call `super()` w/out args

**Review by simahawk — CHANGES_REQUESTED** [review]

**Review by guewen — APPROVED** [review]
Code LGTM (only some commented lines to remove), tested on runbot, works great. Thanks!

**Review by simahawk — APPROVED** [review]

**Review by tschanzt — COMMENTED** [review]
Some of the Headers have the wrong Daterange in it for the licence. But otherwise :+1:

**Comment by grindtildeath** [comment]
@tschanzt Can you approve or add suggestions to correct the date range so we can get this merged ? :pray: :wink:

**Comment by simahawk** [comment]
@tschanzt ping

**Review by tschanzt — APPROVED** [review]

**Comment by benbrich** [comment]
@fclementic2c how about an merge here? Everythings seems allright from over here. Thx!

**Comment by benbrich** [comment]
@fclementic2c sorry for poking you on this one, I have learned from @OSevangelist that you are more into funtional stuff. @tafaRU or @jbeficent could you please be so kind and merge this one? Thanks!

**Pull request closed without merging**